### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by cescoffier**
Provide a source to image for Vert.x

### CODE REVIEW

File: .s2i/bin/assemble [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+#
+# For more information refer to the documentation:
+#	https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md
+#
+
+if [ "$1" = "-h" ]; then
+	# If the 'vertx-s2i' assemble script is executed with '-h' flag,
+	# print the usage.
+	exec /usr/local/sti/usage
+fi
```
Comment (pmlopes) [inline .s2i/bin/assemble:15]: usually the script should stop after the printing the help right?

File: .s2i/bin/assemble [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+DEPLOY_DIR=/opt/openshift
+
+function copy_artifacts() {
+  if [ -d $LOCAL_SOURCE_DIR/$1 ]; then
+    echo "Copying all (fat) JAR artifacts from $LOCAL_SOURCE_DIR/$1 directory into $DEPLOY_DIR for later deployment..."
+    cp -v $LOCAL_SOURCE_DIR/$1/*-fat.jar $DEPLOY_DIR 2> /dev/null
+  fi
+
+	# Copy src/env to deploy directory
```
Comment (pmlopes) [inline .s2i/bin/assemble:45]: messed up indentation?

File: .s2i/bin/assemble [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+# Copy the source for compilation
+cp -ad /tmp/src/* $LOCAL_SOURCE_DIR
+
+# If a pom.xml is present, this is a normal build scenario
+# so run maven.
+if [ -f "$LOCAL_SOURCE_DIR/pom.xml" ]; then
+  echo "Building with maven. $LOCAL_SOURCE_DIR/pom.xml found."
+
+	if [ -z "$OUTPUT_DIR" ]; then
```
Comment (pmlopes) [inline .s2i/bin/assemble:66]: indentiation inconsistent

File: .s2i/bin/assemble [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+  # Copy built artifacts (if any!) from the build/libs (output dir) directory
+  # to the $DEPLOY_DIR directory for later deployment
+  copy_artifacts $OUTPUT_DIR
+
+  # clean up after maven
+  gradle clean
+  if [ -d "$HOME/.m2/repository" ]; then
+    rm -r "$HOME/.m2/repository"
+  fi
```
Comment (pmlopes) [inline .s2i/bin/assemble:142]: what about the gradle caches? `~/.gradle/caches` shouldn't it be clean too?

File: .s2i/bin/run [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+#
+# S2I run script for the 'vertx-sti' image.
+# The run script executes the server that runs your application.
+#
+# For more information see the documentation:
+#	https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md
+#
+
+cd /opt/openshift
+exec java -Dmaven.repo.local=$M2_LOCAL -Dvertx.cacheDirBase=/tmp -Djava.security.egd=file:/dev/./urandom  -jar vertx-app.jar $APP_OPTIONS
```
Comment (pmlopes) [inline .s2i/bin/run]: /dev/urandom is less secure... is this on purpose?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+LABEL io.k8s.description="Platform for building Vert.x applications with maven or gradle" \
+      io.k8s.display-name="Vert.x 3 builder 1.0" \
+      io.openshift.expose-services="8080:http" \
+      io.openshift.tags="builder,maven-3,gradle-2.8,vert.x"
+
+# COPY ./<builder_folder>/ /opt/openshift/
+LABEL io.openshift.s2i.scripts-url=image:///usr/local/sti
+COPY ./.s2i/bin/ /usr/local/s2i
+
+RUN chown -R 1001:1001 /opt/openshift /opt/.m2
```
Comment (pmlopes) [inline Dockerfile:39]: 1001 is a big magical, should it be better to do a adduser? this way you're guaranteed that the uid always exists, say that the base image is updated and the 1001 user is not available anymore...

File: README.adoc [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+BUILDER_ARGS="build -x test"
+OUTPUT_DIR=build/libs
+----
+
+== Test in OpenShift
+
+* First load all the needed resources in a project.
+
+----
+oc create -f https://raw.githubusercontent.com/vert-x3/vertx-openshift-s2i/initial-work/vertx-s2i-all.json
```
Comment (pmlopes) [inline README.adoc]: This will break after merge :) it links to this branch, once merged the link will be broken

File: test/run [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  local sleep_time=1
+  local attempt=1
+  local result=1
+  while [ $attempt -le $max_attempts ]; do
+
+    unamestr=`uname`
+    if [[ "$unamestr" == 'Darwin' ]]; then
+      # Mac os x, with docker-machine
+      echo "Sending GET request to http://$(docker_machine_ip):$(docker_machine_port)/"
+      response_code=$(curl -s -w "%{http_code}" -o /dev/null http://$(docker_machine_ip):$(docker_machine_port)/)
```
Comment (pmlopes) [inline test/run:134]: just a funny think, ubuntu by default does not have curl installed but wget

File: .s2i/bin/assemble (hunk 4 above)
Comment (cescoffier) [inline .s2i/bin/assemble:142]: Yes, it should be also removed.

File: .s2i/bin/run (hunk 5 above)
Comment (cescoffier) [inline .s2i/bin/run]: I took that from the other s2i, I'm not even sure we should set it.

File: Dockerfile (hunk 6 above)
Comment (cescoffier) [inline Dockerfile:39]: Yes, it's a bit magical, but the user is not created by us.

**Pull request merged**